low-level/files/opt/seagate/sspl/setup/resource_map/server.py

```diff
@@ -31,19 +42,200 @@ class ServerMap(ResourceMap):
+            "storage[0]" -> ("storage", "0")
+        """
+        res = re.search(r"(\w+)\[([\d]+)\]|(\w+)", node)
+        inst = res.groups()[1] if res.groups()[1] else "*"
+        node = res.groups()[0] if res.groups()[1] else res.groups()[2]
+        return node, inst
+
+    def get_data_template(self, uid, is_fru: bool):
+        """Get the health data template."""
+        return {
+            "uid": uid,
+            "fru": str(is_fru).lower(),
+            "last_updated": -1,
+            "health": {
+                "status": "",
+                "description": "",
+                "recommendation": "",
+                "specifics": []
+            }
+        }
+
+    def set_health_data(self, obj: dict, status, description=None,
+                        recommendation=None):
+        """Set the given or default health data as per health status."""
+        good_state = (status == "OK")
+        if not description:
+            description = "%s is %sin good health." % (
+                            obj.get("uid"),
+                            '' if good_state else 'not ')
+        if not recommendation:
+            recommendation = 'None' if good_state\
+                             else DEFAULT_RECOMMENDATION
+
+        obj["health"].update({
+            "status": status,
+            "description": description,
+            "recommendation": recommendation
+        })
+
+        obj["last_updated"] = int(time.time())
+
+    def get_sas_hba_info(self):
+        """Return the s SAS-HBA information."""
+        return self.get_sas_ports_info("sas_hba")
+
+    def get_sas_ports_info(self, leaf_node="sas_ports"):
+        """Return the s SAS ports information."""
+        sas_hba_data = []
+        health_data = []
+        sas_hosts = self.sysfs.get_sas_host_list()
+        for i, sas_rid in enumerate(sas_hosts):
+            host_id = SAS_RESOURCE_ID + str(i)
+            r_data = self.get_data_template(host_id, False)
+            sas_ports_dict = self.sysfs.get_phy_negotiated_link_rate()
+            health = []
+            if sas_ports_dict:
+                for rid, rate in sas_ports_dict.items():
+                    r_data["health"]["specifics"].append({
+                        "resource_id": f"{host_id}:{rid.strip()}",
+                        "negotiated_link_rate": rate.strip()
+                    })
+                    if "Gbit".lower() in rate.strip().lower():
+                        health.append("OK")
+
+            health_data.append(health)
+            health_status = "OK" if "OK" in health else "Fault"
+            self.set_health_data(r_data, health_status)
+            if r_data["health"]["specifics"]:
+                sas_hba_data.append(r_data)
+
+        if leaf_node == "sas_hba":
+            return sas_hba_data
+
+        sas_ports_data = []
+        for heath_list in health_data:
+            for i, health in enumerate(heath_list):
+                if i % 4:
+                    continue
```

Comment:
why do we do this?

Reply:
This logic is removed now, we are collecting the port info directly from the /sys/class/sas_port dir